core/shuffle_combiner.hpp

```diff
@@ -0,0 +1,146 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <atomic>
+#include <string>
+#include <vector>
+
+#include "base/serialization.hpp"
+#include "base/thread_support.hpp"
+#include "zmq_helpers.hpp"
+
+namespace husky {
+
+template <typename CollectT>
+class Shuffler {
+   public:
+    Shuffler() {}
+
+    Shuffler(Shuffler&& s) : collection_(s.collection_) { s.collection_ = nullptr; }
+
+    CollectT& storage() {
+        if (collection_ == nullptr)
+            collection_ = new CollectT();
+        return *collection_;
+    }
+
+    virtual ~Shuffler() {
+        if (collection_ != nullptr)
+            delete collection_;
+        collection_ = nullptr;
+    }
+
+   protected:
+    CollectT* collection_ = nullptr;
+};
+
+template <typename CellT>
+class ShuffleCombiner {
+   protected:
+    typedef std::vector<CellT> CollectT;
+
+   public:
+    ShuffleCombiner() {}
+
+    ShuffleCombiner(ShuffleCombiner&& sc) {
+        messages_ = std::move(sc.messages_);
+        publisher_ = sc.publisher_;
+        subscriber_ = sc.subscriber_;
+        num_units_ = sc.num_units_;
+        num_local_workers_ = sc.num_local_workers_;
+        channel_id_ = sc.channel_id_;
+        local_id_ = sc.local_id_;
+        sc.num_units_ = 0;
+        sc.num_local_workers_ = 0;
+        sc.channel_id_ = -1;
+        sc.local_id_ = -1;
+        sc.publisher_ = nullptr;
+        sc.subscriber_ = nullptr;
+    }
+
+    ~ShuffleCombiner() {
+        messages_.clear();
+        delete publisher_;
+        delete subscriber_;
+    }
+
+    void init(int num_global_threads, int num_local_threads, int channel_id, int local_id) {
+        num_units_ = num_global_threads;
+        num_local_workers_ = num_local_threads;
+        channel_id_ = channel_id;
+        local_id_ = local_id;
+        messages_.resize(num_units_);
+        publisher_ = channel_sc_socket_map_[channel_id][local_id_];
+        subscriber_ = channel_sc_socket_map_[channel_id][local_id_ + num_local_workers_];
+    }
+
+    void send_shuffler_buffer() { husky::zmq_send_int32(publisher_, local_id_, ZMQ_NOBLOCK); }
+
+    int access_next() {
+        int worker_id = husky::zmq_recv_int32(subscriber_);
+        return worker_id;
+    }
+
+    CollectT& storage(unsigned idx) { return messages_[idx].storage(); }
+
+    static void init_sockets(int num_local_workers, int channel_id, zmq::context_t& context) {
+        if (channel_sc_socket_map_.find(channel_id) == channel_sc_socket_map_.end()) {

```

Comment:
Okay, maybe we can send two integers one time? the first one is used to identify the channel that should be receiving message and the second integer  is used to specify which worker is available.